File: 2024/day08/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Position struct {
	row, col int
}

func inBounds(r, c, rows, cols int) bool {
	return r >= 0 && r < rows && c >= 0 && c < cols
}

func parseInput(filename string) (int, int, map[rune][]Position, error) {
	file, err := os.Open(filename)
	if err != nil {
		return 0, 0, nil, err
	}
	defer file.Close()

	var grid []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		return 0, 0, nil, err
	}

	rows := len(grid)
	cols := 0
	if rows > 0 {
		cols = len(grid[0])
	}

	// Group antenna positions by frequency
	antennas := make(map[rune][]Position)
	for r, row := range grid {
		for c, ch := range row {
			if ch != '.' {
				antennas[ch] = append(antennas[ch], Position{r, c})
			}
		}
	}

	return rows, cols, antennas, nil
}

func part1(rows, cols int, antennas map[rune][]Position) int {
	antinodes := make(map[Position]struct{})

	for _, positions := range antennas {
		// For each pair of antennas with the same frequency
		for i := 0; i < len(positions); i++ {
			for j := i + 1; j < len(positions); j++ {
				r1, c1 := positions[i].row, positions[i].col
				r2, c2 := positions[j].row, positions[j].col

				// Calculate the two antinodes
				// Antinode beyond antenna 1 (away from antenna 2)
				ar1, ac1 := 2*r1-r2, 2*c1-c2
				// Antinode beyond antenna 2 (away from antenna 1)
				ar2, ac2 := 2*r2-r1, 2*c2-c1

				// Add if within bounds
				if inBounds(ar1, ac1, rows, cols) {
					antinodes[Position{ar1, ac1}] = struct{}{}
				}
				if inBounds(ar2, ac2, rows, cols) {
					antinodes[Position{ar2, ac2}] = struct{}{}
				}
			}
		}
	}

	return len(antinodes)
}

func part2(rows, cols int, antennas map[rune][]Position) int {
	antinodes := make(map[Position]struct{})

	for _, positions := range antennas {
		// For each pair of antennas with the same frequency
		for i := 0; i < len(positions); i++ {
			for j := i + 1; j < len(positions); j++ {
				r1, c1 := positions[i].row, positions[i].col
				r2, c2 := positions[j].row, positions[j].col

				dr, dc := r2-r1, c2-c1

				// Extend in both directions along the line
				// Direction 1: from antenna 1 towards and beyond antenna 2
				r, c := r1, c1
				for inBounds(r, c, rows, cols) {
					antinodes[Position{r, c}] = struct{}{}
					r += dr
					c += dc
				}

				// Direction 2: from antenna 1 away from antenna 2
				r, c = r1-dr, c1-dc
				for inBounds(r, c, rows, cols) {
					antinodes[Position{r, c}] = struct{}{}
					r -= dr
					c -= dc
				}
			}
		}
	}

	return len(antinodes)
}

func main() {
	rows, cols, antennas, err := parseInput("../input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Part 1:", part1(rows, cols, antennas))
	fmt.Println("Part 2:", part2(rows, cols, antennas))
}

File: 2024/day08/java/Solution.java
import java.io.*;
import java.util.*;

public class Solution {
    static class Point {
        final int r;
        final int c;

        Point(int r, int c) {
            this.r = r;
            this.c = c;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Point point = (Point) o;
            return r == point.r && c == point.c;
        }

        @Override
        public int hashCode() {
            return Objects.hash(r, c);
        }
    }

    static class Input {
        final int rows;
        final int cols;
        final Map<Character, List<Point>> antennas;

        Input(int rows, int cols, Map<Character, List<Point>> antennas) {
            this.rows = rows;
            this.cols = cols;
            this.antennas = antennas;
        }
    }

    private static boolean inBounds(int r, int c, int rows, int cols) {
        return r >= 0 && r < rows && c >= 0 && c < cols;
    }

    static Input parseInput(String filename) throws IOException {
        List<String> grid = new ArrayList<>();

        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                grid.add(line);
            }
        }

        int rows = grid.size();
        int cols = rows > 0 ? grid.get(0).length() : 0;

        // Group antenna positions by frequency
        Map<Character, List<Point>> antennas = new HashMap<>();
        for (int r = 0; r < rows; r++) {
            String row = grid.get(r);
            for (int c = 0; c < row.length(); c++) {
                char ch = row.charAt(c);
                if (ch != '.') {
                    antennas.computeIfAbsent(ch, k -> new ArrayList<>()).add(new Point(r, c));
                }
            }
        }

        return new Input(rows, cols, antennas);
    }

    static int part1() throws IOException {
        Input input = parseInput("../input.txt");
        Set<Point> antinodes = new HashSet<>();

        for (Map.Entry<Character, List<Point>> entry : input.antennas.entrySet()) {
            List<Point> positions = entry.getValue();

            // For each pair of antennas with same frequency
            for (int i = 0; i < positions.size(); i++) {
                for (int j = i + 1; j < positions.size(); j++) {
                    Point p1 = positions.get(i);
                    Point p2 = positions.get(j);

                    // Calculate the two antinodes
                    // Antinode beyond antenna 1 (away from antenna 2)
                    int ar1 = 2 * p1.r - p2.r;
                    int ac1 = 2 * p1.c - p2.c;
                    // Antinode beyond antenna 2 (away from antenna 1)
                    int ar2 = 2 * p2.r - p1.r;
                    int ac2 = 2 * p2.c - p1.c;

                    // Add if within bounds
                    if (inBounds(ar1, ac1, input.rows, input.cols)) {
                        antinodes.add(new Point(ar1, ac1));
                    }
                    if (inBounds(ar2, ac2, input.rows, input.cols)) {
                        antinodes.add(new Point(ar2, ac2));
                    }
                }
            }
        }

        return antinodes.size();
    }

    static int part2() throws IOException {
        Input input = parseInput("../input.txt");
        Set<Point> antinodes = new HashSet<>();

        for (Map.Entry<Character, List<Point>> entry : input.antennas.entrySet()) {
            List<Point> positions = entry.getValue();

            // For each pair of antennas with same frequency
            for (int i = 0; i < positions.size(); i++) {
                for (int j = i + 1; j < positions.size(); j++) {
                    Point p1 = positions.get(i);
                    Point p2 = positions.get(j);

                    int dr = p2.r - p1.r;
                    int dc = p2.c - p1.c;

                    // Extend in both directions along the line
                    // Direction 1: from antenna 1 towards and beyond antenna 2
                    int r = p1.r;
                    int c = p1.c;
                    while (inBounds(r, c, input.rows, input.cols)) {
                        antinodes.add(new Point(r, c));
                        r += dr;
                        c += dc;
                    }

                    // Direction 2: from antenna 1 away from antenna 2
                    r = p1.r - dr;
                    c = p1.c - dc;
                    while (inBounds(r, c, input.rows, input.cols)) {
                        antinodes.add(new Point(r, c));
                        r -= dr;
                        c -= dc;
                    }
                }
            }
        }

        return antinodes.size();
    }

    public static void main(String[] args) {
        try {
            System.out.println("Part 1: " + part1());
            System.out.println("Part 2: " + part2());
        } catch (IOException e) {
            System.err.println("Error reading input file: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
